### ISSUE TITLE
[5.x]: Attribute defined test on Component classes always returns `true`

### ISSUE THREAD

**Issue by bencroker** [craftcms/cms#14934]
### Description

The attribute `is defined` test on objects that extend the `Component` class (or any class that implements `__call()`) always returns `true`.

Possibly caused by the changes introduced by @brandonkelly in https://github.com/craftcms/cms/commit/0e31c6701f528f9c41a009a6943e496083614e67.

### Steps to reproduce

1. Render a template with the following Twig code:

```twig
{{ attribute(craft.app, 'xyz') is defined ? 'Method exists' : 'Method does not exist' }}
```

### Expected behavior

It should output `Method does not exist`.

### Actual behavior

It outputs `Method exists`.

### Craft CMS version

5.1.0

**Comment by brandonkelly** [comment]
My memory is that `is defined` has always been unreliable for object properties/methods, because `twig_get_attribute()` basically just throws in the towel and returns `true` [here](https://github.com/twigphp/Twig/blob/a842d75fed59cdbcbd3a3ad7fb9eb768fc350d58/src/Extension/CoreExtension.php#L1699) when it hasn’t been able to find any matching properties or methods on the object.

I just tested on 4.8.9 (last release before 0e31c6701f528f9c41a009a6943e496083614e67) and even 4.7.4 (last release before updating Twig from 3.4 to 3.8 via aa6fe4e854f7f60a1221e1d5af920ce0e0090578), and in both cases I’m getting the same `Method exists` output.

**Comment by bencroker** [comment]
Can confirm that. It looks like the magic `__call()` method trips it up [here](https://github.com/twigphp/Twig/blob/a842d75fed59cdbcbd3a3ad7fb9eb768fc350d58/src/Extension/CoreExtension.php#L1683). 

Reason I’m asking is that this is making Craft Pest fail [here](https://github.com/markhuot/craft-pest-core/blob/cd7edd65aff8068f5de21e59d5dd9fc31c98671e/stubs/compiled_classes/FactoryFields.twig#L7-L10), but I’ll bring it up to @markhuot. Thanks for taking a look!

**Comment by khalwat** [comment]
So I did a deep dive on this a while ago to figure out what was going on... 
```twig
    {% if craft.mattwilcox is defined %}
        <p>Hi, Matt!</p>
    {% endif %}
```
compiles down to this PHP: 
```php
        // line 18
        if (craft\helpers\Template::attribute($this->env, $this->source, ($context["craft"] ?? null), "mattwilcox", [], "any", true, true)) {
            // line 19
            echo "        <p>Hi, Matt!</p>
```
which ends up calling `getAttribute()` in Twig's `CoreExtension.php` class, and because the `craft` variable object implements the PHP magic method `__call` <https://www.php.net/manual/en/language.oop5.overloading.php#object.call> it falls through to here: <https://github.com/twigphp/Twig/blob/3.x/src/Extension/CoreExtension.php#L1698> and since `$isDefinedTest` = `true` it just returns that it's defined.

TL;DR: because the `craft` object implements the PHP `__call` magic method, it could dynamically return anything as being defined, so it just says it's defined. So literally anything will be "defined" as a property on an object that implements `__call` in Twig

**Comment by brandonkelly** [comment]
Right… it’s a byproduct of Twig trying to be extremely lenient on method syntax, e.g. if you type `obj.foo`, that could be referencing a property `foo`, or a method `foo()`/`getFoo()`. And by extension, if `__call()` exists, maybe `__call()` supports a magic method `foo()`.

I feel like the only time `__call()` should be considered is if it’s explicitly a method call though (`obj.foo()`). If the object wants to allow `obj.foo` as a magic _property_, it can already support that syntax via `__get()`.
